// fis-kernel/lib/config.js
/*
 * fis
 * http://fis.baidu.com/
 */

'use strict';

//默认配置
var DEFALUT_SETTINGS = {
    system : {
        repos : 'http://fis.baidu.com/repos'
    },
    project : {
        charset : 'utf8', //字符集
        md5Length : 7, //md5长度
        md5Connector : '_' //md5分隔符
    }
};

//You can't use merge in util.js
//合并操作，直接覆盖。
function merge(source, target){
    if(typeof source === 'object' && typeof target === 'object'){
        for(var key in target){
            if(target.hasOwnProperty(key)){
                source[key] = merge(source[key], target[key]);
            }
        }
    } else {
        source = target;
    }
    return source;
}

var Config = Object.derive({
    constructor : function(){
        this.init.apply(this, arguments);
    },
    init : function(){
        this.data = {}; //配置信息缓存器
        if(arguments.length > 0){
            this.merge.apply(this, arguments); //如果有传配置信息，则合并到缓存器this.data。
        }
        return this; //返回配置实例
    },
    //根据path返回其配置信息，如：'a.b.c' => this.data['a']['b']['c']
    get : function(path, def){
        var result = this.data || {};
        (path || '').split('.').forEach(function(key){
            if(key && (typeof result !== 'undefined')){
                result = result[key];
            }
        });
        if(typeof result === 'undefined'){
            return def;
        } else {
            return result;
        }
    },
    //根据path、value设置配置信息。
    set : function(path, value){
        //如：a.set({...})，直接覆盖this.data。
        if(typeof value === 'undefined'){
            this.data = path;
        } else { //递归设置配置信息。如：'a.b.c' => this.data['a']['b']['c'] = value
            path = String(path || '').trim();
            if(path){
                var paths = path.split('.'),
                    last = paths.pop(),
                    data = this.data || {};
                paths.forEach(function(key){
                    var type = typeof data[key]; //只接受对象或未定义属性
                    if(type === 'object'){
                        data = data[key];
                    } else if(type === 'undefined'){
                        data = data[key] = {};
                    } else {
                        fis.log.error('forbidden to set property[' + key + '] of [' + type + '] data');
                    }
                });
                data[last] = value;
            }
        }
        return this;
    },
    //删除配置，如：'a.b.c' => delete this.data['a']['b']['c']
    del : function(path){
        path = String(path || '').trim();
        if(path){
            var paths = path.split('.'),
                data = this.data,
                last = paths.pop(), key;
            for(var i = 0, len = paths.length; i < len; i++){
                key = paths[i];
                if(typeof data[key] === 'object'){
                    data = data[key];
                } else { //如果不是对象则直接中断遍历，返回实例。
                    return this;
                }
            }
            if(typeof data[last] !== 'undefined'){
                delete data[last];
            }
        }
        return this;
    },
    //将配置信息合并到缓存器this.data。
    merge : function(){
        var self = this;
        [].slice.call(arguments).forEach(function(arg){
            if(typeof arg === 'object'){
                merge(self.data, arg);
            } else {
                fis.log.warning('unable to merge data[' + arg + '].');
            }
        });
        return this;
    },
    //加载配置插件模块，并返回实例，实现链式调用。
    require : function(name){
        fis.require('config', name);
        return this;
    }
});

//返回由Config创建的实例，并合并默认配置DEFAULT_SETTINGS。
module.exports = (new Config).init(DEFALUT_SETTINGS);
//向上面创建的实例添加其构造器Config引用。
module.exports.Config = Config;
